### DISCUSSION TITLE
Tasmota thermostat stops working after some time

### DISCUSSION THREAD

**Discussion by dexterbot80** [arendst/Tasmota#18657] [Q&A]
Hello,
I installed the Tasmota thermostat on the Shelly Uni device and it works but after a while it stops and the temperature drops below the set one.
Can someone help me to make it work so as to keep the temperature in the rooms constant?
I post here the settings I made...
`Rule1 ON Power1#boot DO Backlog sensorinputset 1;controllermodeset 2;thermostatmodeset 1;temptargetset 22;timemaxactionset 3600;timeminactionset 2;timepicycleset 500 %mem1% ENDON`
`Backlog Rule1 1;Rule1 4`
And after some time exactly what you see in the picture happens - although the temperature is set to 22 degrees, the thermostat is turned off and the temperature in the room has dropped to 17.9 degrees.
![WhatsApp Image 2023-05-15 at 10 21 25](https://github.com/arendst/Tasmota/assets/47325794/f3b47dbd-8c4c-46eb-9002-c44a624144ca)
Can someone tell me what I'm doing wrong here?

Thank you very much

**Comment by Wintespe** [comment]
The issue #16424 contains an extensive discussion on this problem, good luck

**Reply by dexterbot80** [reply]
Thanks for the reply.
In my case, the thermostat activates an electric resistance in the floor and thus it has to keep the temperature in the room constant.
To be honest, this thermostat seems very complex to me and I think I just caught my ears :)
It would be much easier if someone would help me with the thermostat settings for the underfloor heating.
I don't understand why it stops - I don't understand what I'm doing wrong here

Thanks

**Comment by Wintespe** [comment]
In my opinion, you are not doing anything wrong with the Thermastat controller. This controller is complex and has problems especially with the predictive ramp-up mode (ControllerModeSet 2). There it comes to the problems described in #16424.

For this reason I use the PID controller https://tasmota.github.io/docs/PID-Control/.
There I have not observed such problems.

**Reply by dexterbot80** [reply]
Hello, Thank you for your answer, Can you help me to configure this PID controller for Shelly Uni device and DS18B20 sensor?
Thanks again

**Comment by Wintespe** [comment]
First you should compile the Tasmoata software with the PID controller enabled.
I have not yet used a Shelly Uni device and can therefore not say anything about the configuration.

Some parameters like the cycletime ( #define TIMEPROP_CYCLETIMES ) have to be adjusted in the source code.

Instructions for setting a PID controller can you find here:
https://tasmota.github.io/docs/PID-Control/
https://blog.clanlaw.org.uk/pid-loop-tuning.html

I derived the following table for myself from the source code of the PID controller:
<html>

<body lang="de-DE" link="#000080" vlink="#800000" dir="ltr">


Command | Def | Description
-- | -- | --
PidPv | - | This Is the current actual 			process value.
PidSp | 19.5 | Setpoint value. This is the 			process value that the process is aiming for.
PidPb | 5 | Proportional band in process 			units (eg degrees). This controls the gain of the loop and is the 			range of process value over which the power output will go from 0 			to full power. The units are that of the process and setpoint, so 			for example in a heating application it might be set to 1.5 			degrees. 			Set to zero 			to implement a 2-step controller.
PidTi | 1800 | This is a setting for the 			integral time, in seconds. It represents the time constant of the 			integration effect. The larger the value the slower the integral 			effect will be. Obviously the slower the process is the larger 			this should be. For example for a domestic room heated by 			convection radiators a setting of one hour might be appropriate 			(in seconds). To disable the integral effect set this to a large 			number,
PidTd | 15 | This is a setting for the 			derivative time, in seconds. It represents the time constant of 			the derivative effect.The larger the value the greater will be the 			derivative effect. Typically this will be set to somewhat less 			than 25% of the integral setting, once the integral has been 			adjusted to the optimum value. To disable the derivative effect 			set this to 0. When initially tuning a loop it is often sensible 			to start with derivative zero and wind it in once other parameters 			have been setup.
PidInitint | 0.5 | This is an initial value 			which is used to preset the integrated error value when the flow 			is deployed in order to assist in homing in on the setpoint the 			first time. It should be set to an estimate of what the power 			requirement might be in order to maintain the process at the 			setpoint. For example for a domestic room heating application it 			might be set to 0.2 indicating that 20% of the available power 			might be required to maintain the setpoint. The value is of no 			consequence apart from device restart.
PidMaxInterval | 300 | This is 			the maximum time in seconds that is expected between samples. It 			is provided to cope with unusual situations such as a faulty 			sensor that might prevent the node from being supplied with a 			process value. If no new process value is received for 			PidMaxInterval 			then the power is set to the value defined by 			 PidManualPower. 			Set to zerro 			if not used. In this case the controller 			is always 			activated.
PidDSmooth | 3 | IIn 			situations where the process sensor has limited resolution (e.g. 			DS18B20), using the derivative can be problematic when the process 			is changing slowly and the steps in the value cause spikes in the 			derivative. To reduce the effect PidDSmoothr 			can be set to apply a filter to the derivative term. It has been 			found that with the DS18B20 a value of 3 can be beneficial here, 			effectively providing a low pass filter for the derivative at 1/3 			the derivative time. This feature may also be useful if the 			process value is particularly noisy. The smaller the value the 			greater the filtering effect but the more it will reduce the 			effectiveness of the derivative. A value of zero disables this 			feature.
PidAuto | 1 | Auto mode 1 or 0 (for 			manual).This can be used to enable or disable the control 			(1=enable, auto mode, 0=disabled, manual mode). When in manual 			mode the output is set the value defined for PidManualPower
PidManualPower | 0 | Power output when in manual 			mode or fallback mode if too long elapses between process values.
PidUpdateSecs | 0 | Specifies 			the 			interval time in seconds  to 			execute the pid algorithm. 			If 			PidUpdateSecs = 0 the algorithm isonly 			triggerd each time a new temperature value is received. This time 			is specified by TELE_PERIOD. 			A 			value 			of 15 seconds may well be 			appropriate for a domestic room heating application.


</body>
</html>>

**Reply by dexterbot80** [reply]
Thank you very much teacher, I will try to implement this function.
I am very curious how the thermostat works with these settings. I use an electric resistance for heating the floor.

**Comment by dexterbot80** [comment]
> First you should compile the Tasmoata software with the PID controller enabled. I have not yet used a Shelly Uni device and can therefore not say anything about the configuration.
> 
> Some parameters like the cycletime ( #define TIMEPROP_CYCLETIMES ) have to be adjusted in the source code.
> 
> Instructions for setting a PID controller can you find here: https://tasmota.github.io/docs/PID-Control/ https://blog.clanlaw.org.uk/pid-loop-tuning.html
> 
> I derived the following table for myself from the source code of the PID controller:
> 
> Command	Def	Description
> PidPv	-	This Is the current actual process value.
> PidSp	19.5	Setpoint value. This is the process value that the process is aiming for.
> PidPb	5	Proportional band in process units (eg degrees). This controls the gain of the loop and is the range of process value over which the power output will go from 0 to full power. The units are that of the process and setpoint, so for example in a heating application it might be set to 1.5 degrees. Set to zero to implement a 2-step controller.
> PidTi	1800	This is a setting for the integral time, in seconds. It represents the time constant of the integration effect. The larger the value the slower the integral effect will be. Obviously the slower the process is the larger this should be. For example for a domestic room heated by convection radiators a setting of one hour might be appropriate (in seconds). To disable the integral effect set this to a large number,
> PidTd	15	This is a setting for the derivative time, in seconds. It represents the time constant of the derivative effect.The larger the value the greater will be the derivative effect. Typically this will be set to somewhat less than 25% of the integral setting, once the integral has been adjusted to the optimum value. To disable the derivative effect set this to 0. When initially tuning a loop it is often sensible to start with derivative zero and wind it in once other parameters have been setup.
> PidInitint	0.5	This is an initial value which is used to preset the integrated error value when the flow is deployed in order to assist in homing in on the setpoint the first time. It should be set to an estimate of what the power requirement might be in order to maintain the process at the setpoint. For example for a domestic room heating application it might be set to 0.2 indicating that 20% of the available power might be required to maintain the setpoint. The value is of no consequence apart from device restart.
> PidMaxInterval	300	This is the maximum time in seconds that is expected between samples. It is provided to cope with unusual situations such as a faulty sensor that might prevent the node from being supplied with a process value. If no new process value is received for PidMaxInterval then the power is set to the value defined by PidManualPower. Set to zerro if not used. In this case the controller is always activated.
> PidDSmooth	3	IIn situations where the process sensor has limited resolution (e.g. DS18B20), using the derivative can be problematic when the process is changing slowly and the steps in the value cause spikes in the derivative. To reduce the effect PidDSmoothr can be set to apply a filter to the derivative term. It has been found that with the DS18B20 a value of 3 can be beneficial here, effectively providing a low pass filter for the derivative at 1/3 the derivative time. This feature may also be useful if the process value is particularly noisy. The smaller the value the greater the filtering effect but the more it will reduce the effectiveness of the derivative. A value of zero disables this feature.
> PidAuto	1	Auto mode 1 or 0 (for manual).This can be used to enable or disable the control (1=enable, auto mode, 0=disabled, manual mode). When in manual mode the output is set the value defined for PidManualPower
> PidManualPower	0	Power output when in manual mode or fallback mode if too long elapses between process values.
> PidUpdateSecs	0	Specifies the interval time in seconds to execute the pid algorithm. If PidUpdateSecs = 0 the algorithm isonly triggerd each time a new temperature value is received. This time is specified by TELE_PERIOD. A value of 15 seconds may well be appropriate for a domestic room heating application.>

Hello @Wintespe, I tried to compile Tasmota master 13.2.0 with pid function enabled but i received this errors.....
Can you help me with that?
Thank you

<img width="1145" alt="Screenshot 2023-10-27 082834" src="https://github.com/arendst/Tasmota/assets/47325794/74474213-4189-4f48-9da4-2aeb4ec25a11">

**Reply by dexterbot80** [reply]
Hello, I managed to install the PID controller but I don't like how it works, I think I'm doing something wrong here.
For example, the temperature in the room is 22 degrees Celsius and I want to raise it to 23. I set the setpoint to 23, but the thermostat is not permanently ON until it reaches 23 degrees. It keeps stopping and starting again... It keeps doing this even if the temperature reaches 23 degrees...
It can be set so that it starts only if the temperature is lower than the setpoint and stops as soon as the temperature reaches the setpoint
I'm not interested in other temperature curves and other functions - I just want it to stay on as long as the temperature in the room is below the set one and to turn off as soon as the temperature in the room reaches the set value.
Can someone help me with the settings for a simple thermostat?
Thank you

**Reply by Wintespe** [reply]
The thermostat is behaving absolutely correctly. To regulate a temperature deviation, a thermostat must switch on and off close to the setpoint. How else would this work?

**Reply by Wintespe** [reply]
Please read about the function of a temperature controller. There is quite good information available on the internet

**Reply by dexterbot80** [reply]
<img width="1270" alt="Screenshot 2024-01-16 184309" src="https://github.com/arendst/Tasmota/assets/47325794/2de4f3af-435e-4969-b25f-a237a8ed0079">

![WhatsApp Image 2024-01-16 at 18 28 37_26b715f6](https://github.com/arendst/Tasmota/assets/47325794/f191a4bb-3008-4392-8ff7-94eee1f000f0)


i do not understand this - setpoint is set to 23 degrees and thermostat is OFF

**Reply by barbudor** [reply]
Then you do not understand what is PID regulation
The principle is to control on/off so the setpoint is reached gently without overshoot
Because of latency between action (on/off on the heater) and reaction (room temp), if the heater was continuously on until 23 is reached, you would likely have a temperature oscillating between 24 and 22

If that's what you want, just use rules instead of a PID regulator

**Reply by dexterbot80** [reply]
hello @barbudor, I appreciate that you want to help me. I really want to control the temperature as accurately as possible, but I don't understand why the thermostat behaves this way.
I don't understand why if I set the setpoint to 23 degrees, the thermostat stops and starts at least 200 times until the temperature reaches 23 degrees.
​I would like to keep the temperature at a fixed 23 degrees if possible, but now I have 21.9 degrees in the room and I set the setpoint to 23. The thermostat stops and starts 200 times until it reaches 23 degrees. I would understand if it would turn ON OFF to maintain a constant 23 degrees, but I don't understand why it continuously turns ON OFF until it reaches 23 degrees. That's why I say that maybe I didn't set something right here.

**Reply by barbudor** [reply]
If you want theory https://en.m.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller

If you don't, just use rules instead of PID

**Reply by dexterbot80** [reply]
I also tried with rules @barbudor  but I encountered another problem - I actually posted it above.
I don't know if the bugs have been solved in the meantime, but it would freeze at some point and had to be reset to works again.....

**Reply by Wintespe** [reply]
A simplified illustration:
Let's assume a set temperature of 23 °C but it is cooler in the room. For example, the radiator has a surface temperature of 70 °C. The temperature sensor is somewhere in the room. The temperature rises and rises until the temperature sensor detects 23 °C and then the heating is switched off. But the radiator is still hot. The temperature continues to rise for a while, causing the temperature to overshoot.  To prevent this, the radiator must be switched off beforehand.  A PID controller is used to achieve this. The parameters must be set correctly for the application.

https://blog.clanlaw.org.uk/pid-loop-tuning.html

**Reply by barbudor** [reply]
Typo
> To prevent this, the radiator must be switched **OFF** beforehand

**Reply by Wintespe** [reply]
Correct. Corrected

**Reply by dexterbot80** [reply]
<img width="1266" alt="image" src="https://github.com/arendst/Tasmota/assets/47325794/a0e40cc4-82c2-4f09-94f6-cdbc1c1d5794">

Hmmm,
And does it have to be off for that long?
I have a floor heating system here and if it works like this, the actuators don't even open and close them again :)

**Reply by Wintespe** [reply]
You can also operate the PID as a two-step controller (like in the rules). To do this, set
Proportional band to 0. (PidPb = 0)
In this case it could look like this

![image](https://github.com/arendst/Tasmota/assets/121858068/74d23935-28dd-4db3-81a7-737c1d0c1549)

**Reply by Wintespe** [reply]
As the temperature control (PidPv - this Is the current actual process value) does not change, it looks as if the heating is not switched on. That is a bit strange. Is the hardware OK?

Why did you set the intergral part so high ( PidTi  ) :  switched off  ?

**Reply by dexterbot80** [reply]
Yes, I tried to set PidPb 0 but now it doesn't stop when it reaches the setpoint :)
This is my setings:
```
Rule1 ON Power1#boot DO Backlog PidPb 0; PidTi 2000; PidTd 5; PidDSmooth 3; PidAuto 1; PidUpdateSecs 10 %mem1% ENDON
Backlog Rule1 1;Rule1 4
```
As I said - I have an underfloor heating system that reacts extremely slowly to both heating and cooling - it needs about 15 minutes to increase the temperature by 0.1 degree
For example, now I have the setpoint at 22 and the room temperature has risen to 22.3. It will go up because the thermostat is still ON

**Reply by Wintespe** [reply]
Please follow the instructions here:
https://blog.clanlaw.org.uk/pid-loop-tuning.html
